Automatically detect MTU of network

tgraf opened the pull request:
Introduces a device MTU to guarantee ability to receive of arbitrarily big
frames. Existing MTU is renamed to RouteMTU and used for all configured routes.

Provide both the route and device MTU via the API so all plugins can use the
MTU configured via the cilium-agent.

Use the MTU of the device of which the default route points to as an indication
of the network MTU. The user can overwrite this using a new --mtu=INT option.

Example:
```
level=info msg="Detected MTU 9001"
level=info msg="Adding route" command="ip route add 0.0.0.0/0 via 10.2.2.172 mtu 8951 dev cilium" netns=cilium-health
level=info msg="Adding route" command="ip -6 route add ::/0 via f00d::a02:200:0:815b mtu 8951 dev cilium" netns=cilium-health
```

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4687)
<!-- Reviewable:end -->

ianvernon commented on Documentation/kubernetes/install.rst:
```diff
@@ -248,16 +248,14 @@ environment variables can be specified in the `DaemonSet` file like this:
     env:
-      - name: "MTU"
-        value: "8950"
+      - name: "CNU_CONF_NAME"
```
CNU --> CNI?

aanm commented on common/plugins/ipam.go:
```diff
@@ -112,7 +111,7 @@ func IPv4Routes(addr *models.NodeAddressing, linkMTU int) ([]Route, error) {
 			Prefix:  node.IPv4DefaultRoute,
 			Nexthop: &ip,
-			MTU:     linkMTU - mtu.TunnelOverhead,
+			MTU:     linkMTU,
```
There's no tunnel overhead anymore?

tgraf replied:
It's abstracted in the mtu package:
```
// GetRouteMTU returns the MTU to be used on the network. When running in
// tunneling mode, this will have tunnel overhead accounted for.
func GetRouteMTU() int {
        return TunnelMTU
}
```

aanm commented on daemon/main.go:
```diff
@@ -399,6 +400,8 @@ func init() {
 		"nat46-range", node.DefaultNAT46Prefix, "IPv6 prefix to map IPv4 addresses to")
 	flags.BoolVar(&masquerade,
 		"masquerade", true, "Masquerade packets from endpoints leaving the host")
+	flags.IntVar(&option.Config.MTU,
```
`flags.IntVar` -> `flags.UIntVar`

tgraf replied:
Do you really care about this? It requires to touch almost every single use of mtu in the code. Do you just care about the user entering negative values? uint is not much better either, the user can still enter invalid values and they will be rejected when we try to apply.

aanm replied:
> Do you just care about the user entering negative values?

Yes, that was the only reason

tgraf replied:
OK, I'll add a validation check in `pkg/option/config.go`, that is a much simpler fix for this

joestringer replied:
Yeah, I started down this track with http://github.com/cilium/mtu-update and realised it's a pain to interact with low-level libraries if you enforce via typed uint rather than just checking against negative values.

aanm commented on pkg/mtu/detect.go:
```diff
@@ -0,0 +1,69 @@
+)
+
+func autoDetect() (int, error) {
+	ip, _, err := net.ParseCIDR(externalProbe)
```
Why not `net.ParseIP`?

aanm commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -61,6 +61,7 @@ cilium-agent
       --logstash-agent string                       Logstash agent address (default "127.0.0.1:8080")
       --logstash-probe-timer uint32                 Logstash probe timer (seconds) (default 10)
       --masquerade                                  Masquerade packets from endpoints leaving the host (default true)
+      --mtu int                                     Overwrite auto-detected MTU of underlying network (default 1500)
```
I got really confused. The default value will not be "1500", it might depend on the user interface's MTU

tgraf replied:
It will be correct when the user actually runs `--help`. This is only the auto generated documentation. I find it very useful that a user can run `cilium-agent --help` and see the value that will be automatically derived.

aanm replied:
Somehow, for some reason, I thought this would be the actual output of the `--help`

aanm commented on Documentation/kubernetes/quickinstall.rst:
```diff
@@ -66,15 +66,6 @@ chapter.
       [adjust the etcd address]
 
 
```
Should we get rid of this file? These instructions seem error prone. Either we should remove it or add the etcd configuration guide here.

tgraf replied:
Definitely out of scope of this PR

aanm replied:
I know, I was just asking if we should delete it or not.

tgraf replied:
TBH, I think it's the best instructions we have right now. The other instructions are so verbose that nobody is following them.

tgraf committed:
doc: Remove obsolete MTU sections

Now that we automatically detect the MTU, the user no longer needs to take care
of this.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
mtu: Introduce device MTU concept

Introduces a device MTU to guarantee ability to receive of arbitrarily big
frames. Existing MTU is renamed to RouteMTU and used for all configured routes.

Provide both the route and device MTU via the API so all plugins can use the
MTU configured via the cilium-agent.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
mtu: Automatically detect MTU of underlying network

Use the MTU of the device of which the default route points to as an indication
of the network MTU. The user can overwrite this using a new --mtu=INT option.

Example:
```
level=info msg="Detected MTU 9001"
level=info msg="Adding route" command="ip route add 0.0.0.0/0 via 10.2.2.172 mtu 8951 dev cilium" netns=cilium-health
level=info msg="Adding route" command="ip -6 route add ::/0 via f00d::a02:200:0:815b mtu 8951 dev cilium" netns=cilium-health
```

Signed-off-by: Thomas Graf <thomas@cilium.io>
